### DISCUSSION TITLE
How can one log and catch all uncaught Nuxt errors?

### DISCUSSION THREAD

**Discussion by gplusdotgr** [nuxt/nuxt#9186] [Questions (Nuxt 2)]
I'm a little lost with error handling and Nuxt, between 
- errorMiddleware
- errorCapture
- Vue.config.errorHandler

The description on the docs is rather...`cryptic`. 
```
Before adding Nuxt error middleware, useful to add your own middleware before using Nuxt's. See the Sentry module for more info.
```
(sentences don't usually start with "before")

Do you mind adding a short explanation on the 3 error handling hooks/handlers and for which case each one should be used for Nuxt?

Additionally, looking at the Sentry module? https://github.com/nuxt-community/sentry-module/blob/v4.0.3/lib/module.js 
 - there's plenty of stuff happening there - what exactly are we looking for?

In my specific case, my stdout is caught by Datadog, so I would only need to log in the right format the error - alongside with the trace. I tried doing that but nothing was logged. (the error was a JS template error)

**Comment by MentalGear** [comment]
Hi @gkatsanos, Did you find a way to handle/log errors in nuxt at a global level and could you share it?

**Reply by gplusdotgr** [reply]
Hey @MentalGear . Sorry for the slow reply! 
Here's my strategy for managing error logging. 
1) I created a custom plugin for logging as other solutions were too complex / not really necessary. It looks a bit like that: 
```js
// plugins/logger.js
const tracer = require('dd-trace')
const formats = require('dd-trace/ext/formats')

export default (_, inject) => {
  function logger(level, payload, message) {
    const span = tracer.scope().active()
    const time = new Date().toISOString()
    const record = { time, level, ...payload, message }

    if (span) {
      tracer.inject(span.context(), formats.LOG, record) // straight from the Datadog documentation
    }

    console.log(record)
  }

  inject('logger', logger)
}

```

as you know logger is now available on the app level with `this.$logger` or `app.$logger` depending on where you access it (Client Side / Server Side)

2) I added an `errorCaptured` hook in my `default.vue` layout https://vuejs.org/v2/api/#errorCaptured 
This catches all template- related issues from all child components. It also gives you as arguments many useful things, including the `vm` which tells you which component threw the error ( `vm.$options._componentTag` worked for me ). There's elso the error and some information which you could inspect to see if they're useful for you. Be sure to `return false` so the error isnt propagated further. 

3) I personally added interceptors in my axios module : https://axios.nuxtjs.org/helpers/#interceptors which catch and log problematic requests on the API level. Again using my custom logger plugin. 

4) Be sure to not ommit setting the default Nuxt logger to send JSON instead of strings...

```js
// nuxt.config.js file
consola.setReporters([
  {
    log: (e) => {
      process.stdout.write(`${JSON.stringify(...e.args)}\n`)
    },
  },
])
```
 the above is critical - otherwise you'll see a bunch of single-line entries for each log in Datadog.. 

https://www.patreon.com/gplusdotgr
